runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java

```diff
@@ -1296,6 +1302,35 @@ private StreamingPubsubIOWriteOverrideFactory(DataflowRunner runner) {
   }
 
   @VisibleForTesting
+  static class StreamingShardedWriteFactory<T>
+      implements PTransformOverrideFactory<PCollection<T>, PDone, WriteFiles<T>> {
+    DataflowPipelineWorkerPoolOptions options;
+
+    StreamingShardedWriteFactory(PipelineOptions options) {
+      this.options = options.as(DataflowPipelineWorkerPoolOptions.class);
+    }
+
+    @Override
+    public PTransformReplacement<PCollection<T>, PDone> getReplacementTransform(
+        AppliedPTransform<PCollection<T>, PDone, WriteFiles<T>> transform) {
+      // By default, if numShards is not set WriteFiles will produce one file per bundle. In
+      // streaming, there are large numbers of small bundles, resulting in many tiny files.
+      // Instead we pick max workers * 2 to ensure full parallelism, but prevent too-many files.
+      // (current_num_workers * 2 might be a better choice, but that value is not easily available
+      // today).
+      return PTransformReplacement.of(
+            PTransformReplacements.getSingletonMainInput(transform),
+            transform.getTransform().withNumShards(options.getMaxNumWorkers() * 2));
```

Comment:
I think the test error is because getMaxNumWorkers() can return 0, and in that case you return a transform whose numShards is again 0, which again matches this override (overrides must converge into a state where no transform matches any override anymore).